### DISCUSSION TITLE
Weird behaviour with AsyncClient

### DISCUSSION THREAD

**Discussion by piermarcob** [encode/httpx#2263] [Potential Issue]
Dear all, I hope I find you good.

I'm facing a weird behaviour with the `AsyncClient`. My scenario is the following

- Frontend `POST`ing to some backend API.
- Backend accepting the request and returning a 301, with a `Location` value set to a remote website: `https://...`.
- The `AsyncClient` sends a `GET` to a local resource, which clearly does not exist: the request becomes `http://localhost:8000/https://...`  and the response code is 404 from my local backend.

It seems the client cannot recognise whether the `Location` value refers to a local or a remote path, therefore a local one is always applied.

The desired outcome would be to actually redirect the client to the remote website `https://...`, with a response status code 200.

I hope i made myself explained.

Let me know what you think.

Regards.

**Comment by lovelydinosaur** [comment]
So I took a look at your description [of this on Gitter too](https://gitter.im/encode/community?at=62a4e77d4aa6c31dca489cba).

> The AsyncClient sends a GET to a local resource, which clearly does not exist: the request becomes http://localhost:8000/https://... and the response code is 404 from my local backend.

So I don't think this description quite matches what's actually happening. I double checked and the URL redirection is correctly formed, and isn't to the composite URL. But I'll describe this more carefully just in case there's still any confusion.

You've got a test client, which is making requests directly to a web app rather than actual network requests.

The initial request is "to" `http://localhost:8000/`. The host doesn't actually matter here because you're sending the request to your Python ASGI app, not out to the network.

The redirected request is "to" `https://www.google.com/`. But again, the host doesn't actually matter here because you're sending requests to your ASGI app, not out to the network.

**Reply by piermarcob** [reply]
Hi Tom, thanks for the explanation, i was indeed missing things out.
The question now becomes: How should i setup my tests such that i can actually verify the redirection and that the location received in the headers gives a 200 OK back?

**Reply by lovelydinosaur** [reply]
> How should i setup my tests such that i can actually verify the redirection and that the location received in the headers gives a 200 OK back?

I'll start answering that by using an explicit transport for the `ASGITransport`, rather than the `AsyncClient(app=...)` shortcut.

Here's an equivalent of what you're currently using...

```python
transport = httpx.ASGITransport(app=app)
async with httpx.AsyncClient(transport=transport) as client:
    ...
```

Typically if I was writing a test like this, I wouldn't want the third party requests to *actually* send out to the network. You generally want to mock out that service at the test layer.

Here's the type of pattern I'd probs be looking for...

```python
transport = httpx.ASGITransport(app=app)
mock_google_transport = httpx.ASGITransport(app=mock_service)
mounts = {
    'https://testserver': transport,
    'https://google.com': mock_google_transport
}

async with httpx.AsyncClient(mounts=mounts) as client:
    ...
```

**Reply by piermarcob** [reply]
Okay, so I think at this point i'll stick with just testing that a redirect came back from my backend and ensuring the redirect location matches the one i want my clients to be redirected to.
